In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Display all columns
pd.set_option('display.max_columns', None)

# Display more rows
pd.set_option('display.max_rows', 100)

In [2]:
matches = pd.read_csv("C:/Users/anujm/OneDrive/Documents/Desktop/CricketIQ/data/raw/matches.csv")
deliveries = pd.read_csv("C:/Users/anujm/OneDrive/Documents/Desktop/CricketIQ/data/raw/deliveries.csv")

In [3]:
matches.head()
matches.info()
matches.shape
matches.columns
matches.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095 entries, 0 to 1094
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               1095 non-null   int64  
 1   season           1095 non-null   object 
 2   city             1044 non-null   object 
 3   date             1095 non-null   object 
 4   match_type       1095 non-null   object 
 5   player_of_match  1090 non-null   object 
 6   venue            1095 non-null   object 
 7   team1            1095 non-null   object 
 8   team2            1095 non-null   object 
 9   toss_winner      1095 non-null   object 
 10  toss_decision    1095 non-null   object 
 11  winner           1090 non-null   object 
 12  result           1095 non-null   object 
 13  result_margin    1076 non-null   float64
 14  target_runs      1092 non-null   float64
 15  target_overs     1092 non-null   float64
 16  super_over       1095 non-null   object 
 17  method        

id                    0
season                0
city                 51
date                  0
match_type            0
player_of_match       5
venue                 0
team1                 0
team2                 0
toss_winner           0
toss_decision         0
winner                5
result                0
result_margin        19
target_runs           3
target_overs          3
super_over            0
method             1074
umpire1               0
umpire2               0
dtype: int64

In [4]:
deliveries.head()
deliveries.info()
deliveries.shape
deliveries.columns
deliveries.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 260920 entries, 0 to 260919
Data columns (total 17 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          260920 non-null  int64 
 1   inning            260920 non-null  int64 
 2   batting_team      260920 non-null  object
 3   bowling_team      260920 non-null  object
 4   over              260920 non-null  int64 
 5   ball              260920 non-null  int64 
 6   batter            260920 non-null  object
 7   bowler            260920 non-null  object
 8   non_striker       260920 non-null  object
 9   batsman_runs      260920 non-null  int64 
 10  extra_runs        260920 non-null  int64 
 11  total_runs        260920 non-null  int64 
 12  extras_type       14125 non-null   object
 13  is_wicket         260920 non-null  int64 
 14  player_dismissed  12950 non-null   object
 15  dismissal_kind    12950 non-null   object
 16  fielder           9354 non-null    obj

match_id                 0
inning                   0
batting_team             0
bowling_team             0
over                     0
ball                     0
batter                   0
bowler                   0
non_striker              0
batsman_runs             0
extra_runs               0
total_runs               0
extras_type         246795
is_wicket                0
player_dismissed    247970
dismissal_kind      247970
fielder             251566
dtype: int64

## Initial Observations

### Matches Dataset
- The dataset contains 1,095 IPL matches with 20 columns.
- A few columns have missing values, such as city, winner, player_of_match, result_margin, target_runs, target_overs, and method.
- The 'method' column has many missing values because it is only populated for DLS-affected matches.
- Most missing values appear to be due to match conditions rather than data quality issues.

### Deliveries Dataset
- The dataset contains 260,920 ball-by-ball records across IPL matches.
- Most columns have complete data.
- Missing values in extras_type, player_dismissed, dismissal_kind, and fielder are expected because they are only applicable in specific situations.
- This dataset will be the primary source for player-level, team-level, and match-level analytics.

# Data Cleaning

In [5]:
print("Duplicate rows in Matches:", matches.duplicated().sum())
print("Duplicate rows in Deliveries:", deliveries.duplicated().sum())

Duplicate rows in Matches: 0
Duplicate rows in Deliveries: 0


In [6]:
matches.dtypes

id                   int64
season              object
city                object
date                object
match_type          object
player_of_match     object
venue               object
team1               object
team2               object
toss_winner         object
toss_decision       object
winner              object
result              object
result_margin      float64
target_runs        float64
target_overs       float64
super_over          object
method              object
umpire1             object
umpire2             object
dtype: object

In [7]:
deliveries.dtypes

match_id             int64
inning               int64
batting_team        object
bowling_team        object
over                 int64
ball                 int64
batter              object
bowler              object
non_striker         object
batsman_runs         int64
extra_runs           int64
total_runs           int64
extras_type         object
is_wicket            int64
player_dismissed    object
dismissal_kind      object
fielder             object
dtype: object

In [8]:
matches["date"] = pd.to_datetime(matches["date"])

In [9]:
matches.dtypes

id                          int64
season                     object
city                       object
date               datetime64[ns]
match_type                 object
player_of_match            object
venue                      object
team1                      object
team2                      object
toss_winner                object
toss_decision              object
winner                     object
result                     object
result_margin             float64
target_runs               float64
target_overs              float64
super_over                 object
method                     object
umpire1                    object
umpire2                    object
dtype: object

### Cleaning Step 1: Convert Date Column

**Reason:**
The `date` column is currently stored as an object (string). Converting it to the datetime format enables time-based analysis, feature engineering, and easier filtering of matches by year, month, or day.

**Action Taken:**
Converted the `date` column using `pd.to_datetime()`.

**Expected Result:**
The `date` column will have the data type `datetime64[ns]`.

In [10]:
matches.isnull().sum()

id                    0
season                0
city                 51
date                  0
match_type            0
player_of_match       5
venue                 0
team1                 0
team2                 0
toss_winner           0
toss_decision         0
winner                5
result                0
result_margin        19
target_runs           3
target_overs          3
super_over            0
method             1074
umpire1               0
umpire2               0
dtype: int64

In [11]:
matches["team1"].unique()

array(['Royal Challengers Bangalore', 'Kings XI Punjab',
       'Delhi Daredevils', 'Mumbai Indians', 'Kolkata Knight Riders',
       'Rajasthan Royals', 'Deccan Chargers', 'Chennai Super Kings',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Sunrisers Hyderabad',
       'Gujarat Lions', 'Rising Pune Supergiants',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Lucknow Super Giants', 'Gujarat Titans',
       'Royal Challengers Bengaluru'], dtype=object)

In [12]:
matches["team2"].unique()

array(['Kolkata Knight Riders', 'Chennai Super Kings', 'Rajasthan Royals',
       'Royal Challengers Bangalore', 'Deccan Chargers',
       'Kings XI Punjab', 'Delhi Daredevils', 'Mumbai Indians',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Sunrisers Hyderabad',
       'Rising Pune Supergiants', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Gujarat Titans', 'Lucknow Super Giants',
       'Royal Challengers Bengaluru'], dtype=object)

In [13]:
matches["season"].unique()

array(['2007/08', '2009', '2009/10', '2011', '2012', '2013', '2014',
       '2015', '2016', '2017', '2018', '2019', '2020/21', '2021', '2022',
       '2023', '2024'], dtype=object)

In [14]:
seaoson_name_mapping={
    '2007/08':'2008',
    '2009/10':'2010',
    '2020/21':'2020'
}

In [15]:
sorted(matches["team1"].unique())
sorted(deliveries["batting_team"].unique())

['Chennai Super Kings',
 'Deccan Chargers',
 'Delhi Capitals',
 'Delhi Daredevils',
 'Gujarat Lions',
 'Gujarat Titans',
 'Kings XI Punjab',
 'Kochi Tuskers Kerala',
 'Kolkata Knight Riders',
 'Lucknow Super Giants',
 'Mumbai Indians',
 'Pune Warriors',
 'Punjab Kings',
 'Rajasthan Royals',
 'Rising Pune Supergiant',
 'Rising Pune Supergiants',
 'Royal Challengers Bangalore',
 'Royal Challengers Bengaluru',
 'Sunrisers Hyderabad']

In [16]:
team_name_mapping = {
    "Delhi Daredevils": "Delhi Capitals",
    "Kings XI Punjab": "Punjab Kings",
    "Royal Challengers Bangalore": "Royal Challengers Bengaluru",
    "Rising Pune Supergiant": "Rising Pune Supergiants"
}

In [17]:
team_columns = ["team1", "team2", "toss_winner", "winner"]

for col in team_columns:
    matches[col] = matches[col].replace(team_name_mapping)

In [18]:
deliveries["batting_team"] = deliveries["batting_team"].replace(team_name_mapping)
deliveries["bowling_team"] = deliveries["bowling_team"].replace(team_name_mapping)

In [19]:
sorted(matches["team1"].unique())

['Chennai Super Kings',
 'Deccan Chargers',
 'Delhi Capitals',
 'Gujarat Lions',
 'Gujarat Titans',
 'Kochi Tuskers Kerala',
 'Kolkata Knight Riders',
 'Lucknow Super Giants',
 'Mumbai Indians',
 'Pune Warriors',
 'Punjab Kings',
 'Rajasthan Royals',
 'Rising Pune Supergiants',
 'Royal Challengers Bengaluru',
 'Sunrisers Hyderabad']

In [20]:
venue_name_mapping = {
    # Pune
    "Subrata Roy Sahara Stadium": "Maharashtra Cricket Association Stadium, Pune",

    # Delhi
    "Feroz Shah Kotla": "Arun Jaitley Stadium, Delhi",

    # Bengaluru
    "M Chinnaswamy Stadium": "M Chinnaswamy Stadium, Bengaluru",

    # Hyderabad
    "Rajiv Gandhi International Stadium, Uppal":
    "Rajiv Gandhi International Stadium, Uppal, Hyderabad",

    # Kolkata
    "Eden Gardens, Kolkata": "Eden Gardens",

    # Mohali
    "Punjab Cricket Association Stadium, Mohali":
    "Punjab Cricket Association IS Bindra Stadium",

    "Punjab Cricket Association IS Bindra Stadium, Mohali, Chandigarh":
    "Punjab Cricket Association IS Bindra Stadium",

    # Brabourne
    "Brabourne Stadium, Mumbai": "Brabourne Stadium"
}

matches["venue"] = matches["venue"].replace(venue_name_mapping)

## Data Cleaning Summary

The following cleaning steps were performed:

- Checked for duplicate records.
- Converted the `date` column to datetime format.
- Investigated missing values instead of filling them blindly.
- Standardized franchise names across all datasets.
- Verified data consistency after cleaning.

In [21]:
matches.to_csv("C:/Users/anujm/OneDrive/Documents/Desktop/CricketIQ/data/processed/matches_cleaned.csv", index=False)
deliveries.to_csv("C:/Users/anujm/OneDrive/Documents/Desktop/CricketIQ/data/processed/deliveries_cleaned.csv", index=False)